docs/blueprints/EP038.rst

```diff
@@ -0,0 +1,169 @@
+:EP: 38
+:Title: Action Pacing
+:Authors:
+    - David Ramirez <davramir@fiu.edu>;
+:Issued Date: 2024-05-16
+:Status: Pending
+:Type: Standards Track
+
+*********************
+EP038 - Action Pacing
+*********************
+
+########
+Abstract
+########
+
+This document is intended on a guide for both NApp Developers
+and Network Operators for how to implement and use Pacing within Kytos.
+Pacing is intended as a mechanism to control
+how often certain actions will be executed.
+This differs from rate limiting, as the action is not cancelled,
+and instead the caller is temporarily paused until the action
+is allowed to be executed again.
+For an action to be paced, it needs to be instrumented
+by developers to support pacing.
+
+##########
+Motivation
+##########
+
+The purpose of adding in pacing to Kytos is
+help maintain system stability by preventing
+excessive throughput.
+It could occur that sending too many flow mods
+or too many DB requests may result in
+connection failures, and therefore system
+instability.
+To prevent such scenarios,
+pacing of actions which are sensitive to high
+throuhput should be implemented.
+
+#############
+Specification
+#############
+
+This section describes how to set the pace of actions,
+intended for network operators,
+and how to add pacing to an action,
+intended for developers.
+
+Setting A Pace
+==============
+
+For network operators,
+setting the pace of an action can be done through
+a NApp's ``action_paces`` setting.
+The ``action_paces`` setting is a dictionary,
+with keys representing the action to be paced,
+and values being the rate to pace them to.
+The pacing value is a string formatted as
+``{max_concurrent}/{time_period}{time_unit}``,
+where ``max_concurrent`` is the max amount of times
+the action is allowed to be executed
+within the ``time_period`` using the given ``time_unit``.
+Here's an example:
+
+.. code-block:: python
+
+    action_paces = {
+        "send_flow_mod": "50/2seconds",
+    }
+
+How the ``time_period`` is interpreted is controlled by the pacing strategy.
+Pacing strategy can be set using the ``pacing_strategy`` option in ``kytos.conf``.
```

Comment:
Can we not have any pacer settings on `kytos.conf` in this iteration? If you could also parametrize the strategy to be passed with `action_paces`. Let's allow whoever will instantiate the pacer to also pass the strategy. Same idea as the reason why we're having this on NApps settings and not on `kytos.conf`. Or let me know if you had other reasons why.

Reply:
The only problem with having the strategies be a per action parameter is having to have multiple strategy objects, but that's not that big a deal. I can switch this over to being part of `action_paces`.

Reply:
Right. Let's go for it then.